### PR TITLE
add mechanism for configuring system image builds

### PR DESCRIPTION
**Pull request by JeffBezanson**
This adds the option to pass a filename of configuration settings when building the Core/compiler system image (from `base/compiler/compiler.jl`). This makes it easier to build different flavors of images, for example it can replace the hack that PackageCompiler uses to edit the list of included stdlibs, and makes it easy to change knobs you might want like max_methods.

~~Will need corresponding changes to PackageCompiler.~~

### CODE REVIEW

**Comment by Seelengrab** [comment]
Kind of tangential, but having full-on Julia source code as the configuration file format seems VERY dangerous, especially if people start sharing these. Would using a TOML be an option instead? That would also have the advantage of making the configuration file easier to document in terms of what the compiler actually looks for in the configuration options.

**Comment by KristofferC** [comment]
It would at least be good to have support for setting these from e.g. `Make.user` (similar to how one sets `JULIA_PRECOMPILE`, `JULIA_CPU_TARGET` etc).

**Comment by JeffBezanson** [comment]
TOML would be good, we just don't yet have a parser available at this stage. It will be awkward, since we don't even have arrays or dicts at this stage.

The target user for this is code like PackageCompiler that is building custom system images. I don't expect anyone to change these settings just when building julia with Make.user. For comparison, currently PackageCompiler literally rewrites the file `sysimg.jl` to make changes like this. So the threat model is like somebody sending you a file and telling you to replace sysimg.jl with it --- obviously that is dangerous and we can't prevent it.

**Comment by Seelengrab** [comment]
> TOML would be good, we just don't yet have a parser available at this stage. It will be awkward, since we don't even have arrays or dicts at this stage.

Yeah, I figured.. It's a bootstrapping thing all over again!

File: base/settings.jl [hunk 1]
```diff
@@ -0,0 +1,11 @@
+baremodule Settings
+using Core: ARGS, include, Intrinsics, memoryrefget, memoryref
+
+MAX_METHODS::Int = 3
+UNOPTIMIZE_THROW_BLOCKS::Bool = true
+INCLUDE_STDLIBS::String = "FileWatching,Libdl,Artifacts,SHA,Sockets,LinearAlgebra,Random"
+
+if Intrinsics.slt_int(1, getfield(getfield(ARGS, :size), 1))
+    include(Settings, memoryrefget(memoryref(ARGS.ref, 2, false), :not_atomic, false))
+end
```
Comment (topolarity) [inline base/settings.jl:10]: What's this magic here?

File: base/settings.jl [hunk 2]
```diff
@@ -0,0 +1,11 @@
+baremodule Settings
+using Core: ARGS, include, Intrinsics, memoryrefget, memoryref
+
+MAX_METHODS::Int = 3
+UNOPTIMIZE_THROW_BLOCKS::Bool = true
+INCLUDE_STDLIBS::String = "FileWatching,Libdl,Artifacts,SHA,Sockets,LinearAlgebra,Random"
```
Comment (vchuravy) [inline base/settings.jl:6]: Can we keep the comments from the previous file.

File: base/settings.jl (hunk 1 above)
Comment (oscardssmith) [reply base/settings.jl:10]: this is an array getindex that is hard-coded presumably due to bootstraping issues.

Comment (oscardssmith) [reply base/settings.jl:10]: specifically, this is
```
if 1 <= ARGS.size[1]
    include(Settings, ARGS[2])
end
```

**Comment by JeffBezanson** [comment]
OK I have a better implementation:

- Core.Compiler and Base each have their own settings, since they are separate libraries. You can pass `--settings` to either.
- In the process, DRY the buildroot argument to the Base image build
- No more unreadable intrinsics code
- Backwards compatible with existing PackageCompiler

**Comment by JeffBezanson** [comment]
Decided to call this "buildsettings" to clarify they are build-time.

**Pull request merged**